### PR TITLE
96 filename handlebars

### PR DESCRIPTION
**Pull request by bmuenzenmeyer**
WIP implementation of https://github.com/transitive-bullshit/create-react-library/issues/96

Feedback is welcome!

## New Feature
Template file names are now processed via Handlebars, with `name` supplied as the only data

Example: `{{shortName}}.js` 

the following helpers are baked in (assuming name is `my-test-library`)

```
{{camelCase shortName}} // myTestLibrary
{{kebabCase shortName}} // my-test-library
{{pascalCase shortName}} // MyTestLibrary
{{snakeCase shortName}} // my_test_libraray
```

these helpers work within all files or within filenames

`{{pascalCase shortName}}.js`

## Initial Testing

I've added a file called `{{shortName.txt}}` to `template/`. I want CI to run here as opposed to my machine since it has generated local false positives with previous work.

I have this working in my own project like so:

image

`{{pascalCase shortName}}.js`

``` js
import React from 'react'
import { array, bool, func, node, object, oneOfType, string } from 'prop-types'

export class {{ pascalCase shortName }} extends React.Component {
  render() {
    return '{{pascalCase shortName}}'
  }
}

{{ pascalCase shortName }}.defaultProps = {
}

{{ pascalCase shortName }}.propTypes = {
}

export default {{ pascalCase shortName }}

```

generates (with name as `component-button-test4`) `ComponentButtonTest4.js`

``` js
import React from 'react'
import { array, bool, func, node, object, oneOfType, string } from 'prop-types'

export class ComponentButtonTest4 extends React.Component {
  render() {
    return 'ComponentButtonTest4'
  }
}

ComponentButtonTest4.defaultProps = {
}

ComponentButtonTest4.propTypes = {
}

export default ComponentButtonTest4

```

## Specific Considerations

1. If this lands, how should we document it in the `README`, if at all?
1. The initial data is limited to only shortName. I did this since that seems like the only thing you'd pass to the fileName parsing.
1. I included only helpers that I conceivably would want to use to name files with. Is this an okay default set? 
1. ~~I have not tested `@scoped` packages yet.  Is `name` the correct key?  I see `shortName` also available.~~ Updated to use `shortName`

### CODE REVIEW

**Comment by bmuenzenmeyer** [comment]
@transitive-bullshit 
CI seems to have stalled for an unknown reason. hoping it can get resolved. 🤷‍♂️

File: lib/registerHelpers.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+const kebabCase = require('lodash.kebabcase')
+const mixin = require('lodash.mixin')
+const snakeCase = require('lodash.snakecase')
+const upperFirst = require('lodash.upperfirst')
+
+// best solution for now https://github.com/lodash/lodash/pull/942
+const m = mixin({'pascalCase': flow(camelCase, upperFirst)})
+
+module.exports = handlebars => {
+  handlebars.registerHelper({
```
Comment (bmuenzenmeyer) [inline lib/registerHelpers.js]: I got some internal feedback about the casing of these helpers. I see inconsistent use across the Handlebars documentation, so any direction from the community is helpful here.

**Comment by transitive-bullshit** [comment]
@bmuenzenmeyer this is definitely interesting 😄 Thank you for your contributions!

In general, custom templates are not something that have received that much feedback from the community. I'd love to hear more about your concrete target use case from #96.

Thanks again!

**Comment by bmuenzenmeyer** [comment]
@transitive-bullshit no thank you. This is a rock-solid solution for our use case. Myself and @zachwolf are eager to help where we can. 

Our plan is to incorporate CRL into a larger packaged monorepo (think a fork of create-react-app that we've extended with subcommands). We like CRL's custom template as a means to create a consistent structure for our components, and hope to get contributions from others within the org using our custom template. The shipped `Example` component is just extra work a user has to alter post-generation, either deleting or renaming. Why not create the proper name right off the bat? 😁  

I am working on switching my implementation to use `shortName` instead of `name`, and have identified a couple problems with my CI tests that I am working through. Even so, testing is slow going on my machine due to the thoroughness (a good thing!)

Will keep you informed.

**Comment by bmuenzenmeyer** [comment]
Committing a pivot to `shortName` to this PR. The CI is very resource-intensive to run on my machine, so using Travis feels more efficient to me

**Comment by bmuenzenmeyer** [comment]
Not yet sure what's wrong with the last two tests - they seem unrelated to my changes and more to do with create-react-app's jest tests and roll up. Will try to reproduce locally, but my first go was unsuccessful

**Comment by bmuenzenmeyer** [comment]
@transitive-bullshit not sure yet what the remaining CI test fails mean. I am not seeing this locally.

**Comment by bmuenzenmeyer** [comment]
Took another look at this - not sure why jest cli is not being found...

**Comment by transitive-bullshit** [comment]
I definitely appreciate the work you did on this, but maintenance time is always an issue on OSS projects and this complicates things a bit, I'm going to go ahead and close this.

Feel free to reopen if you feel strongly.

Thanks!

**Pull request closed without merging**